cranelift/codegen/src/isa/riscv64/inst/emit.rs

```diff
@@ -2893,6 +2893,40 @@ impl MachInstEmit for Inst {
                 .emit(&[], sink, emit_info, state);
                 sink.bind_label(label_done);
             }
+            &Inst::ElfTlsGetAddr { ref symbol, rd } => {
+                let rd = allocs.next_writable(rd);
+                assert_eq!(a0(), rd.to_reg());
+                sink.add_reloc(Reloc::RiscvTlsGdHi20, symbol, 0);
+                Inst::Auipc {
+                    rd: rd,
+                    imm: Imm20::from_bits(0),
+                }
+                .emit(&[], sink, emit_info, state);
+
+                sink.add_reloc(Reloc::RiscvPCRelLo12I, symbol, 0);
```

Comment:
So I just tried this and the linker gave an error. When I tried comparing it with what LLVM produces I found that this relocation needs to have a label pointing to the previous instruction as target: https://godbolt.org/z/8P31Gcc7Y @cfallin how should an api for allowing backends to produce this kind of labels look like?

Reply:
@bjorn3  In that case I think one reloc maybe be  more reasonable like `ElfTlsGD`,`ElfTlsGD` means 
~~~
.LBB0_1:
        auipc   a0, %tls_gd_pcrel_hi(example::FOO)
        addi    a0, a0, %pcrel_lo(.LBB0_1)
~~~
the whole bunch of things.
`ElfTlsGD+0` means the label and tls_gd_pcrel_hi, `ElfTlsGD + 4` means the pcrel_lo.

Reply:
https://github.com/llvm/llvm-project/blob/9c626d4a0d4d75c976c9a7c9093ea885fcbd98d5/lld/ELF/InputSection.cpp#L542-L580

It doesn't seem to work that way it seems. For R_RISCV_PCREL_LO12_{I,S} the target is always a label pointing to a place with a R_RISCV_PCREL_HI20, R_RISCV_GOT_HI20, R_RISCV_TLS_GD_HI20 or R_RISCV_TLS_GOT_HI20. The actual symbol for which the R_RISCV_PCREL_LO12_{I,S} works on is taken from this relocation, rather than from the R_RISCV_PCREL_LO12_{I,S} itself. In other words it uses `example::FOO` as symbol rather than `.LBB0_1`.

Reply:
> @cfallin how should an api for allowing backends to produce this kind of labels look like?

In this case the `MachInst` should generate both instructions I think (`auipc` and `addi`); that way the label use is internal to the `MachInst`. Anything more complex than that needs a lot of additional plumbing that we don't have...

Reply:
Yeah, the label needs to be exported as entry in the ELF symbol table.